client/lib/interval/runner.js

```diff
@@ -0,0 +1,97 @@
+import { fromJS } from 'immutable';
+
+export const EVERY_SECOND = 1000;
+export const EVERY_FIVE_SECONDS = 5 * 1000;
+export const EVERY_TEN_SECONDS = 10 * 1000;
+export const EVERY_THIRTY_SECONDS = 30 * 1000;
+export const EVERY_MINUTE = 60 * 1000;
+
+const initialState = fromJS( {
+	nextId: 1,
+	periodTimers: {
+		EVERY_SECOND: null,
+		EVERY_FIVE_SECONDS: null,
+		EVERY_TEN_SECONDS: null,
+		EVERY_THIRTY_SECONDS: null,
+		EVERY_MINUTE: null
+	},
+	actions: []
+} );
+let state = initialState;
+
+const increment = a => a + 1;
+const push = item => list => list.push( item );
+const pull = id => list => list.filterNot( o => o.get( 'id' ) === id );
+
+export const resetForTesting = () => {
+	state
+		.get( 'periodTimers' )
+		.forEach( clearTimeout );
+
+	state = initialState;
+};
+
+export function add( period, onTick ) {
+	const id = state.get( 'nextId' );
+
+	storeNewAction( { id, period, onTick } );
+	scheduleNextRun();
+
+	return id;
+}
+
+function storeNewAction( { id, period, onTick } ) {
+	state = state
+		.update( 'actions', push( fromJS( { id, period, onTick } ) ) )
```

Comment:
OK, so `push` is a function that takes an item and returns a function that takes a list as an argument, which will then append the item to the list. So `state.update` must take its first argument `actions` to pull an element from the immutable object `state`, and then pass that into the second argument, which is a function. So I think it's my lack of familiarity with `immutable-js` (and working with immutable objects in general) that's biting me here. But what bugs me is that I can't find an explanation of `.update()` in the [immutable docs](http://facebook.github.io/immutable-js/docs/).